File: sdks/python/src/echo_agent_sdk/permission_rule_values.py
from __future__ import annotations

from enum import Enum


class RuleSource(str, Enum):
    DEFAULT = "default"
    LOCAL_SETTINGS = "localSettings"
    PROJECT_SETTINGS = "projectSettings"
    USER_SETTINGS = "userSettings"
    MANAGED = "managed"
    CLI_ARG = "cliArg"
    SESSION = "session"

    @classmethod
    def parse(cls, value: str) -> RuleSource:
        if not isinstance(value, str):
            raise TypeError("rule source must be text")
        aliases = {
            "default": cls.DEFAULT,
            "localSettings": cls.LOCAL_SETTINGS,
            "local_settings": cls.LOCAL_SETTINGS,
            "projectSettings": cls.PROJECT_SETTINGS,
            "project_settings": cls.PROJECT_SETTINGS,
            "userSettings": cls.USER_SETTINGS,
            "user_settings": cls.USER_SETTINGS,
            "manual": cls.USER_SETTINGS,
            "managed": cls.MANAGED,
            "cliArg": cls.CLI_ARG,
            "cli_arg": cls.CLI_ARG,
            "session": cls.SESSION,
        }
        try:
            return aliases[value]
        except KeyError as error:
            raise ValueError(f"unknown permission rule source: {value}") from error

    def __str__(self) -> str:
        return self.value

File: sdks/typescript/src/permission_rule_values.ts
/** Permission rule source values projected without owning evaluation order. */
export const RuleSource = Object.freeze({
  Default: "default",
  LocalSettings: "localSettings",
  ProjectSettings: "projectSettings",
  UserSettings: "userSettings",
  Managed: "managed",
  CliArg: "cliArg",
  Session: "session",
} as const);

export type RuleSource = (typeof RuleSource)[keyof typeof RuleSource];

export function ruleSourceParse(value: string): RuleSource {
  if (typeof value !== "string") throw new TypeError("rule source must be text");
  const aliases: Readonly<Record<string, RuleSource>> = {
    default: RuleSource.Default,
    localSettings: RuleSource.LocalSettings,
    local_settings: RuleSource.LocalSettings,
    projectSettings: RuleSource.ProjectSettings,
    project_settings: RuleSource.ProjectSettings,
    userSettings: RuleSource.UserSettings,
    user_settings: RuleSource.UserSettings,
    manual: RuleSource.UserSettings,
    managed: RuleSource.Managed,
    cliArg: RuleSource.CliArg,
    cli_arg: RuleSource.CliArg,
    session: RuleSource.Session,
  };
  const source = aliases[value];
  if (source === undefined) throw new Error(`unknown permission rule source: ${value}`);
  return source;
}
